rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,194 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__TOPIC_STATISTICS__SUBSCRIBER_TOPIC_STATISTICS_HPP_
+#define RCLCPP__TOPIC_STATISTICS__SUBSCRIBER_TOPIC_STATISTICS_HPP_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcl/time.h"
+#include "rclcpp/time.hpp"
+#include "rclcpp/publisher.hpp"
+#include "rclcpp/timer.hpp"
+
+#include "metrics_statistics_msgs/msg/metrics_message.hpp"
+
+#include "libstatistics_collector/collector/generate_statistics_message.hpp"
+#include "libstatistics_collector/topic_statistics_collector/constants.hpp"
+#include "libstatistics_collector/topic_statistics_collector/received_message_period.hpp"
+#include "libstatistics_collector/moving_average_statistics/types.hpp"
+
+
+namespace rclcpp
+{
+namespace topic_statistics
+{
+
+using libstatistics_collector::collector::GenerateStatisticMessage;
+using metrics_statistics_msgs::msg::MetricsMessage;
+using libstatistics_collector::moving_average_statistics::StatisticData;
+
+/**
+ * Class used to collect, measure, and publish topic statistics data. Current statistics
+ * supported for subscribers are received message age and received message period.
+ *
+ * @tparam CallbackMessageT the subscribed message type
+ */
+template<typename CallbackMessageT>
+class SubscriberTopicStatistics
+{
+  using topic_stats_collector =
+    libstatistics_collector::topic_statistics_collector::TopicStatisticsCollector<
+    CallbackMessageT>;
+  using received_message_period =
+    libstatistics_collector::topic_statistics_collector::ReceivedMessagePeriodCollector<
+    CallbackMessageT>;
+
+public:
+  /// Construct a SubcriberTopicStatistics object.
+  /**
+   * This object wraps utilities, defined in libstatistics_collector, to collect,
+   * measure, and publish topic statistics data.
+   *
+   * @param node_name the name of the node, which created this instance, in order to denote
+   * topic source
+   * @param publisher instance constructed by the node in order to publish statistics data
+   */
+  SubscriberTopicStatistics(
+    const std::string & node_name,
+    const rclcpp::Publisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr & publisher)
+  : node_name_(node_name),
+    publisher_(std::move(publisher))
+  {
+    // TODO(dbbonnie): ros-tooling/aws-roadmap/issues/226, received message age
+
+    auto rmp = std::make_unique<received_message_period>();
+    rmp->Start();
+    subscriber_statistics_collectors_.emplace_back(std::move(rmp));
+
+    window_start_ = rclcpp::Time(GetCurrentNanosecondsSinceEpoch());
+  }
+
+  virtual ~SubscriberTopicStatistics()
+  {
+    TearDown();
```

Comment:
Should the user call explicitly or not? If yes, we don't need this call in the dtor (or we should log a warning at least). If no, `TearDown` should be private.

Reply:
Will make private.